Cloud Terraform provider beta (#11)
redpanda-data/cloud-docs: modules/develop/pages/terraform-provider.adoc

You can use the Redpanda Terraform provider to create and manage the following resources in Redpanda Cloud:

* ACLs
* Clusters 
* Networks
* Resource groups
* Topics
* Users

**Reviewers: Can we add roles to the list?**

== Prerequisites

* Install Terraform using the https://learn.hashicorp.com/tutorials/terraform/install-cli[Terraform documentation^].
* Log in to https://cloud.redpanda.com[Redpanda Cloud^], navigate to the *Organization IAM - Service account* page, and create a service account. The service account acts as an OAuth 2.0 client, and you will use its credentials (client ID and client secret) to authenticate to the Cloud API. 

== Limitations

The following functionality is supported in the Cloud API but not in the Redpanda Terraform provider: 

* Secrets
* Kafka Connect

== Use the provider

. To install the Redpanda provider, copy and paste the following code into your Terraform configuration file:
+
```
terraform {
  required_providers {
    redpanda = {
      source  = "redpanda-data/redpanda"

>       version = "~> 0.8"

Reviewer: Should we maintain this version somewhere, or even fetch it dinamically? cc @JakeSCahill

Reviewer 2: @micheleRP We should now bump this to 0.9.0

= Redpanda Terraform Provider
:description: Learn how to use the Redpanda Terraform provider to create and manage resources in Redpanda Cloud.
:page-cloud: true
:page-beta: true

The https://registry.terraform.io/providers/redpanda-data/redpanda/latest[Redpanda Terraform provider^] interfaces with the Cloud API, so you can define your Redpanda infrastructure in Terraform configuration files. You can use the Terraform provider to create and manage the following resources in Redpanda Cloud:

* ACLs
* Clusters 
* Networks
* Resource groups
* Topics
* Users

== Prerequisites

Install Terraform following the https://learn.hashicorp.com/tutorials/terraform/install-cli[Terraform documentation^].

== Limitations

*?? TO REVIEWERS: Can you please confirm that the provider is supported on Serverless and Azure? And what functionality is supported in Cloud API but not provider, like Connectors?*

> 

Reviewer 3: Yes serverless, azure no, connectors is correct. Secrets are related to connectors and also not supported. As far as I understand we've implemented the other stuff

== Use the provider

. To install the Redpanda provider, copy and paste the following code into your Terraform configuration file:
+
```
terraform {
  required_providers {
    redpanda = {
      source  = "redpanda-data/redpanda"
      version = "~> 0.5.1"
    }
  }
}
```

. Initialize Terraform:
+
```
terraform init
```

. For the provider to authenticate against Redpanda Cloud, enter your `client_id` and `client_secret` credentials in your Terraform configuration file or set them as environment variables:
+
[tabs]
======
Static credentials::
+
--
```
provider "redpanda" {
  client_id      = "<client_id>"
  client_secret  = "<client_secret>"
}
```
--
Environment variables::
+
--
```
REDPANDA_CLIENT_ID=<client_id>
REDPANDA_CLIENT_SECRET=<client_secret>
```
--
======

> . For sample configurations, see the https://registry.terraform.io/providers/redpanda-data/redpanda/latest/docs[Redpanda Terraform provider docs^]. *NOTE TO REVIEWERS: This linked doc needs update for beta*

Reviewer 3: there is no linked doc?

Author: @gene-redpanda I'm thinking [this doc](https://registry.terraform.io/providers/redpanda-data/redpanda/latest/docs) should be updated from alpha to beta

Reviewer 3: Oh sure, we can do that next release.

:description: Learn how to use the Redpanda Terraform provider to create and manage resources in Redpanda Cloud.
:page-cloud: true
:page-beta: true

A Terraform provider is a plugin that interacts with cloud providers, third-party tools, and other APIs. The https://registry.terraform.io/providers/redpanda-data/redpanda/latest[Redpanda Terraform provider^] interfaces with the Redpanda Cloud API, so you can define your Redpanda infrastructure in Terraform configuration files. 

You can use the Redpanda Terraform provider to create and manage the following resources in Redpanda Cloud:

* ACLs
* Clusters 
* Networks
* Resource groups
* Topics
* Users

== Prerequisites

Install Terraform following the https://learn.hashicorp.com/tutorials/terraform/install-cli[Terraform documentation^].

== Limitations

The following functionality is supported in the Cloud API but not in the Redpanda Terraform provider: 

* Managed connectors
* Secrets

== Use the provider

. To install the Redpanda provider, copy and paste the following code into your Terraform configuration file:
+
```
terraform {
  required_providers {
    redpanda = {
      source  = "redpanda-data/redpanda"

>       version = "~> 0.5.1"

Reviewer 2: ```suggestion
      version = "~> 0.7/0"
```

It looks like the latest version of the Operator is 0.7.0? @gene-redpanda

Reviewer 3: That is the current latest version of the provider, but we are pending a release pretty soon.

Reviewer 2: We should update to at least 0.8.0. We are trying to shoot for potentially a public beta by next week if possible.

Reviewer 4: @gene-redpanda What is the version that should be used here? This content goes out Sept 30th.

Reviewer 3: `version = "~> 0.8"` is what i'd recommend

    }
  }
}
```

. Initialize Terraform:
+
```
terraform init
```

> 

Reviewer 2: Could we link here to how to create a client and how to get the `client_id` and `client_secret`?

Author: @david-yu Should we add a prereq on this page to create a service account in the Redpanda Cloud UI? Then in this step, explain how to retrieve the `client_id` and `client_secret` in the UI?

Reviewer 2: Yes that would work!

Reviewer 3: Worth mentioning that the end user can set REDPANDA_CLIENT_ID and REDPANDA_CLIENT_SECRET in their environment to avoid setting these values in HCL

Author: I've updated the [prereqs & step #3](https://deploy-preview-11--rp-cloud.netlify.app/redpanda-cloud/develop/terraform-provider/#prerequisites). I _think_ with RBAC, the page name will be called *Organization IAM*, but that may change.

. For the provider to authenticate against Redpanda Cloud, enter your `client_id` and `client_secret` credentials in your Terraform configuration file or set them as environment variables:
+
[tabs]
======

> Static credentials::

Reviewer 2: Couuld we place a warning that this is not recommended for production? Typically folks would want to set these using environment variables or retrieve via some KMS system.

Author: @david-yu My understanding is that this whole feature is going out as beta, and beta is not considered production-ready... so I don't think we want an additional "not recommended for production" for this section. If we don't recommend static credentials, should we just remove that option and say to set environment variables?

= Redpanda Terraform Provider
:description: Learn how to use the Redpanda Terraform provider to create and manage resources in Redpanda Cloud.
:page-cloud: true
:page-beta: true

> A Terraform provider is a plugin that interacts with cloud providers, third-party tools, and other APIs. The https://registry.terraform.io/providers/redpanda-data/redpanda/latest[Redpanda Terraform provider^] interfaces with the Redpanda Cloud API, so you can define your Redpanda infrastructure in Terraform configuration files. 

Reviewer 4: ```suggestion
A Terraform provider is a plugin that interacts with cloud providers, third-party tools, and other APIs. The https://registry.terraform.io/providers/redpanda-data/redpanda/latest[Redpanda Terraform provider^] interfaces with the Redpanda Cloud API, enabling you to define your Redpanda infrastructure in Terraform configuration files. 
```

You can use the Redpanda Terraform provider to create and manage the following resources in Redpanda Cloud:

* ACLs
* Clusters 
* Networks
* Resource groups
* Topics
* Users

== Prerequisites

> * Install Terraform following the https://learn.hashicorp.com/tutorials/terraform/install-cli[Terraform documentation^].

Reviewer 4: ```suggestion
* Install Terraform using the https://learn.hashicorp.com/tutorials/terraform/install-cli[Terraform documentation^].
```

* Log in to https://cloud.redpanda.com[Redpanda Cloud^], navigate to the *Organization IAM - Service account* page, and create a service account. The service account acts as an OAuth 2.0 client, and you will use its credentials (client ID and client secret) to authenticate to the Cloud API. 

== Limitations

The following functionality is supported in the Cloud API but not in the Redpanda Terraform provider: 

> * Managed connectors

Reviewer 4: Aren't we calling this Kafka Connect now (instead of Managed Connectors)? Please check with @asimms41.

Author: Thanks @Feediver1: yes, I've corrected this!

+
--
```
provider "redpanda" {
  client_id      = "<client_id>"
  client_secret  = "<client_secret>"
}
```
--
======

. For sample configurations, see the https://registry.terraform.io/providers/redpanda-data/redpanda/latest/docs[Redpanda Terraform provider docs^]. 

=== Use the provider to create a new cluster

For an example of how to use the provider to create a new cluster, see https://github.com/redpanda-data/terraform-provider-redpanda/blob/main/examples/cluster/aws/main.tf[this example^].

=== Use the provider to manage an existing cluster

To manage resources in existing Redpanda Cloud clusters, you must reference the existing cluster using the cluster ID (Redpanda ID). The following example creates a topic in a cluster with ID `foo-bar-baz`. The `redpanda_topic` resource contains a field `cluster_api_url` that references the `data.redpanda_cluster.test.cluster_api_url` data resource. 

```
data "redpanda_cluster" "test" {
  id = "foo-bar-baz"
}

resource "redpanda_topic" "test" {
  name               = "myTopicName"
  partition_count    = 3
  replication_factor = 3
  cluster_api_url    = data.redpanda_cluster.test.cluster_api_url
}
```

=== Additional examples

> To see more ways to use the provider, see the https://github.com/redpanda-data/terraform-provider-redpanda/tree/main/examples[Redpanda Terraform provider examples^].

Reviewer 4: ```suggestion
To view additional ways to use the provider, see the https://github.com/redpanda-data/terraform-provider-redpanda/tree/main/examples[Redpanda Terraform provider examples^].
```

Reviewer 4: @micheleRP Where is the Beta badge?
